README.md
# CMS-370-GCCLIB
Native VM/CMS C Library. Developed by Robert O'Hara, Redmond Washington, April 2009.  Based on code written by Paul Edwards and modified by Dave Wade. Released to the public domain.


changelog.txt
GCCLIB Change LOG
=================

1.0.0 - 15 Apr 2021
F0054 - Updated Build process for CE
        Removed Old files / Old Build Process
F0053 - Removed SYSPROFB EXEC

0.8.5 - 14 Mar 2021
F0052 - Renamed the new CMSdebug() to CMSsetdebug() to avoid the duplicate!

0.8.4 - 14 Mar 2021
F0051 - Added CMSdirectRead() Read from the virtual console, bypassing stack
F0050 - Added GCCLIB [VERSION|DEBUG] argument
          VERSION prints version
          DEBUG prints version, anchor and turns on debug mode
      - Added CMSdebug() to set debug mode (enables memory leak warning)
      - Initialise "scale" local variable, in log() in math.c

0.8.3 - 4 Feb 2021
F0049 - EPLIST Parsing Error

0.8.2 - 15 Dec 2020
F0048 - Code reformatting
        Added -D__CMS__ to CMAKE so that the right #ifdefs work including code
        reformatting
F0047 - Code reformatting
        Added fake CMAKE so the IDEs can show code errors/warnings
        Local Build - Stop MSYS2 rewriting directory paths in the docker container

0.8.1 - 26 Nov 2020
F0046 - FIX - fwrite() crashed when size or count are zero
      - Implemented CMSfunctionDataArray() for binary arguments for calltype 5 calls
F0045 - Theme - make calltype 5 handle binary data
      - Implement CMSreturndata()
      - Change CMSfunction calltype family of functions to return the returned data length
        on success
      - Updated tests
      - Updated local build script (cp2vm370.sh) for MacOS/Linux

0.8.0-  4 Oct 2020
F0044 - Fix for __ZONE when HRC700DK not installed
F0043 - Fix for Calltype 5 allocation (double word not fullword)
      - cmssys.assemble
        __STACKN entry point to call the SENTRIES command provided by HRC406DS
        __ZONE entry point to return the time zone offset from GMT in seconds
      - time.c use __ZONE to get local time
F0042 - Fix for Calltype 5 without a return buffer
        Fix for CMSargc() return type
        Fix for tstarg2 CMSplist() casting
F0041 - REXXSAA RexxVariablePool

0.7.19- 31 August 2020
F0040 - Fix for calltype 5 returns where RC is auto-converted to the return
        string. The integer RC is set to zero as the call itself was successfull
        The return string has the 'correct' return code.
      - Added 0x to the anchor address announcement

0.7.18- 28 August 2020
F0039 - Reoranised header files a bit to *reduce* circular dependancies
      - Changed CMSGetPG() to macro - performance


0.7.17- 11 August 2020
F0038 - Add support for being called via a type 5 call
      - Assist to turn main() return int into a calltype 5 return EVALBLOK string
      - Functions to set return evalblok
      - Functions to access PLIST, EPLIST, argv, argc, callype and func/subroutine values
      - Parameters passed in an array or passed as a variable number of parameters supported
      - Simple Call Macros CMSsimplefuntion() CMSsimpleprocedure()
F0037 - CMSfunction() - call type 5
F0036 - Process Global Memory: CMSPGAll() & CMSGetPG()

0.7.16- 26 July 2020
F0035 - Minor: Explict cast (int to void*) in resmain.c
      - GIT Config to force LF file line endings
      - Changed SYSPROF EXEC to fixed record format (Workaround for a argument
        defect)

0.7.15- 25 July 2020
F0034 - Update to SYSPROF EXEC to initalise BREXX Anchor
F0033 - Include Option for Alternative Stub Build (thanks Bob Bolch)
        See top of MKGCCLIB EXEC and MKGCCMAC EXEC for opton
      - Changed YATA option to read directly from READER. The archive had
        too many records for the file system (an OMG moment!), it shows off the
        new gcclib unix like stream handling, READER is just like any file :-)
F0032 - Remove EPLIST Magic Marker Hack! True fix done by HRC402DS CMS Mod
      - Move to builder tag image for builds
      - Added Argument Test

0.7.14- 22 July 2020
F0031 - Fix to punchc parameters (thanks Bob Bolch)
F0030 - Fix r+ mode variable record files so the last record no
        longer truncated to the existing max record
        Fix fwrite() in text mode needs now processes newlines
        Optimiser turned on again!

0.7.13- 19 July 2020
F0029 - fgethandle() now ignores any lrecl or recfm information
      - REXX Assist: nextrecLen()
      - ftell() error code fix

0.7.12- 14 July 2020
F0028 - REXX ASSISTS:
          fateof() - is at the end of the file
          fgethandle() - if the file is open
          fgetrecs()
          fgetchars()
          append() - the oposite of rewind()
F0027 - Fix SIGTERM handler overwriting User Exit function
F0026 - Automate Stub Build (C based builder MAKELIB)
      - Re-entrant version of CMSSYS (Thanks to Bob Bolch)
      - Re-entrant version of CMSENTRY (Thanks to Bob Bolch)
      - ABORT announces errno / message
      - DLMALLOC announces line number of any panic
      - GCCLIB announces version
      - GCCLIB sets Stub Anchor in NUCON (NUCRSV7)
      - Performance tuned DYNSTAK (Thanks to Bob Bolch)
F0025 - Changed to -O2 optimisation (-O3 seems to have a bug)
      - Changed code to support -Wall -pedantic -ansi builds
      - Remove stdout, stdin, stdout, errno, etc. from client library static globals
      - Removed strtok()'s static variable

0.7.11- 22 June 2020
F0024 - IO Refactoring / Drivers
      - Record cache for random record access
      - Read/Write files
      - Close files on exit
      - fix __wrterm() needed on startup
      - fix fopen() with missing file mode
      - removed notyet()
      - setvbuf()/setbuf() Implemented
      - tmpfile() implemented
      - putc()/putchar()/getc()/getchar() turned into macros
      - freopen() Implemented
      - fread()/fputs()/fgets() no longer just use putc()/getc() - performance
      - IO Test Suite
F0023 - Fix for Issue #12 (CMS System Function should be functions not defines)
F0022 - Fix for Issue #3 (Writing a empty line causes an error)
F0021 - Fix for Issue #1 (ctype.h has c++ style comments)
F0020 - Implement atexit()
F0019 - SIG_DFL, SIG_ERR, SIG_IGN made pointers to non-real functions

time.c
/*********************************************************************/
/*                                                                   */
/*  This Program Written by Paul Edwards.                            */
/*  Released to the Public Domain                                    */
/*                                                                   */
/*********************************************************************/
/*********************************************************************/
/*                                                                   */
/*  time.c - implementation of stuff in time.h                       */
/*                                                                   */
/*********************************************************************/

#include "time.h"
#include "stdarg.h"
#include "stdio.h"
#include "stddef.h"

/* CMS Time Zone function */
#if defined(__CMS__)

int __zone();


#endif

/* pdos and msdos use the same interface most of the time) */
#ifdef __PDOS__
#define __MSDOS__
#endif

#if defined(__MVS__)
#include "mvssupa.h"
#endif

#ifdef __OS2__
#include <os2.h>
#endif
#ifdef __WIN32__

#include <windows.h>

#endif
#ifdef __MSDOS__
#ifdef __WATCOMC__
#define CTYP __cdecl
#else
#define CTYP
#endif
void CTYP __datetime(void *ptr);
#endif

/* scalar date routines    --    public domain by Ray Gardner
** These will work over the range 1-01-01 thru 14699-12-31
** The functions written by Ray are isleap, months_to_days,
** years_to_days, ymd_to_scalar, scalar_to_ymd.
** modified slightly by Paul Edwards
*/

static int isleap(unsigned yr) {
    return yr % 400 == 0 || (yr % 4 == 0 && yr % 100 != 0);
}

static unsigned months_to_days(unsigned month) {
    return (month * 3057 - 3007) / 100;
}

static long years_to_days(unsigned yr) {
    return yr * 365L + yr / 4 - yr / 100 + yr / 400;
}

static long ymd_to_scalar(unsigned yr, unsigned mo, unsigned day) {
    long scalar;

    scalar = day + months_to_days(mo);
    if (mo > 2)                         /* adjust if past February */
        scalar -= isleap(yr) ? 1 : 2;
    yr--;
    scalar += years_to_days(yr);
    return (scalar);
}

static void scalar_to_ymd(long scalar,
                          unsigned *pyr,
                          unsigned *pmo,
                          unsigned *pday) {
    unsigned n;                /* compute inverse of years_to_days() */

    n = (unsigned) ((scalar * 400L) / 146097L);
    while (years_to_days(n) < scalar) {
        n++;
    }
    for (n = (unsigned) ((scalar * 400L) / 146097L); years_to_days(n) < scalar;)
        n++;                          /* 146097 == years_to_days(400) */
    *pyr = n;
    n = (unsigned) (scalar - years_to_days(n - 1));
    if (n > 59) {                       /* adjust if past February */
        n += 2;
        if (isleap(*pyr))
            n -= n > 62 ? 1 : 2;
    }
    *pmo = (n * 100 + 3007) / 3057;  /* inverse of months_to_days() */
    *pday = n - months_to_days(*pmo);
    return;
}

time_t time(time_t *timer) {
    time_t tt;
#ifdef __OS2__
    DATETIME dt;
    APIRET rc;
#endif
#ifdef __WIN32__
    SYSTEMTIME dt;
#endif
#if defined(__MSDOS__)
    struct {
        int year;
        int month;
        int day;
        int hours;
        int minutes;
        int seconds;
        int hundredths;
    } dt;
#endif
#if defined(__MVS__) || defined(__CMS__)
    unsigned int clk[2];
#endif

#ifdef __OS2__
    rc = DosGetDateTime(&dt);
    if (rc != 0)
    {
        tt = (time_t)-1;
    }
    else
#endif
#ifdef __WIN32__
    GetSystemTime(&dt);
    tt = ymd_to_scalar(dt.wYear, dt.wMonth, dt.wDay)
         - ymd_to_scalar(1970, 1, 1);
    tt = tt * 24 + dt.wHour;
    tt = tt * 60 + dt.wMinute;
    tt = tt * 60 + dt.wSecond;
#endif
#ifdef __MSDOS__
    __datetime(&dt);
#endif
#if defined(__MVS__) || defined(__CMS__)
    tt = __getclk(clk);
#elif !defined(__WIN32__)

    {
        tt = ymd_to_scalar(dt.year, dt.month, dt.day)
             - ymd_to_scalar(1970, 1, 1);
        tt = tt * 24 + dt.hours;
        tt = tt * 60 + dt.minutes;
        tt = tt * 60 + dt.seconds;
    }
#endif
    if (timer != NULL) {
        *timer = tt;
    }
    return (tt);
}

clock_t clock(void) {
    return ((clock_t) -1);
}

double difftime(time_t time1, time_t time0) {
    return ((double) (time1 - time0));
}

time_t mktime(struct tm *timeptr) {
    time_t tt;

    if ((timeptr->tm_year < 70) || (timeptr->tm_year > 120)) {
        tt = (time_t) -1;
    } else {
        tt = ymd_to_scalar(timeptr->tm_year + 1900,
                           timeptr->tm_mon + 1,
                           timeptr->tm_mday)
             - ymd_to_scalar(1970, 1, 1);
        tt = tt * 24 + timeptr->tm_hour;
        tt = tt * 60 + timeptr->tm_min;
        tt = tt * 60 + timeptr->tm_sec;
    }
    return (tt);
}

char *asctime(const struct tm *timeptr) {
    static const char wday_name[7][3] = {
            "Sun", "Mon", "Tue", "Wed", "Thu", "Fri", "Sat"
    };
    static const char mon_name[12][3] = {
            "Jan", "Feb", "Mar", "Apr", "May", "Jun",
            "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"
    };
    char *result = GETGCCCRAB()->atresult;

    sprintf(result, "%.3s %.3s%3d %.2d:%.2d:%.2d %d\n",
            wday_name[timeptr->tm_wday],
            mon_name[timeptr->tm_mon],
            timeptr->tm_mday, timeptr->tm_hour,
            timeptr->tm_min, timeptr->tm_sec,
            1900 + timeptr->tm_year);
    return result;
}

char *ctime(const time_t *timer) {
    return (asctime(localtime(timer)));
}

/* dow - written by Paul Edwards, 1993-01-31 */
/* Released to the Public Domain */
/* This routine will work over the range 1-01-01 to 32767-12-31.
   It assumes the current calendar system has always been in
   place in that time.  If you pass 0 or negative years, then
   it produces results on the assumption that there is a year
   0.  It should always produce a value in the range of 0..6
   if a valid month and day have been passed, no matter what
   the year is.  However, it has not been tested for negative
   years, because the results are meaningless anyway.  It is
   mainly to stop people playing silly buggers and causing
   the macro to crash on negative years. */

#define dow(y, m, d) \
  ((((((m)+9)%12+1)<<4)%27 + (d) + 1 + \
  ((y)%400+400) + ((y)%400+400)/4 - ((y)%400+400)/100 + \
  (((m)<=2) ? ( \
  (((((y)%4)==0) && (((y)%100)!=0)) || (((y)%400)==0)) \
  ? 5 : 6) : 0)) % 7)

struct tm *localtime(const time_t *timer) {
#if defined(__CMS__)
    time_t local_timer = *timer + __zone();
    return (gmtime(&local_timer));
#else
    return (gmtime(timer));
#endif
}

struct tm *gmtime(const time_t *timer) {
    struct tm *tms = GETGCCCRAB()->gmtimetm;
    unsigned yr, mo, da;
    unsigned long secs;
    unsigned long days;

    days = *timer / (60L * 60 * 24);
    secs = *timer % (60L * 60 * 24);
    scalar_to_ymd(days + ymd_to_scalar(1970, 1, 1), &yr, &mo, &da);
    tms->tm_year = yr - 1900;
    tms->tm_mon = mo - 1;
    tms->tm_mday = da;
    tms->tm_yday = (int) (ymd_to_scalar(tms->tm_year + 1900, mo, da)
                         - ymd_to_scalar(tms->tm_year + 1900, 1, 1));
    tms->tm_wday = dow(tms->tm_year + 1900, mo, da);
    tms->tm_isdst = -1;
    tms->tm_sec = (int) (secs % 60);
    secs /= 60;
    tms->tm_min = (int) (secs % 60);
    secs /= 60;
    tms->tm_hour = (int) secs;
    return (tms);
}

/*
 * strftime.c
 *
 * implements the iso c function strftime()
 *
 * written 1989-09-06 by jim nutt
 * released into the public domain by jim nutt
 *
 * modified 1989-10-21 by Rob Duff
 *
 * modified 1994-08-26 by Paul Edwards
 */

static const char *aday[] = {
        "Sun", "Mon", "Tue", "Wed", "Thu", "Fri", "Sat"
};

static const char *day[] = {
        "Sunday", "Monday", "Tuesday", "Wednesday",
        "Thursday", "Friday", "Saturday"
};

static const char *amonth[] = {
        "Jan", "Feb", "Mar", "Apr", "May", "Jun",
        "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"
};

static const char *month[] = {
        "January", "February", "March", "April", "May", "June",
        "July", "August", "September", "October", "November", "December"
};

static const char *__tzname[2] = {"" ""};

static void strfmt(char *str, const char *fmt, ...);

/**
 *
 * size_t strftime(char *str,
 *                 size_t maxs,
 *                 const char *fmt,
 *                 const struct tm *t)
 *
 *      this functions acts much like a sprintf for time/date output.
 *      given a pointer to an output buffer, a format string and a
 *      time, it copies the time to the output buffer formatted in
 *      accordance with the format string.  the parameters are used
 *      as follows:
 *
 *          str is a pointer to the output buffer, there should
 *          be at least maxs characters available at the address
 *          pointed to by str.
 *
 *          maxs is the maximum number of characters to be copied
 *          into the output buffer, included the '\0' terminator
 *
 *          fmt is the format string.  a percent sign (%) is used
 *          to indicate that the following character is a special
 *          format character.  the following are valid format
 *          characters:
 *
 *              %A      full weekday name (Monday)
 *              %a      abbreviated weekday name (Mon)
 *              %B      full month name (January)
 *              %b      abbreviated month name (Jan)
 *              %c      standard date and time representation
 *              %d      day-of-month (01-31)
 *              %H      hour (24 hour clock) (00-23)
 *              %I      hour (12 hour clock) (01-12)
 *              %j      day-of-year (001-366)
 *              %M      minute (00-59)
 *              %m      month (01-12)
 *              %p      local equivalent of AM or PM
 *              %S      second (00-59)
 *              %U      week-of-year, first day sunday (00-53)
 *              %W      week-of-year, first day monday (00-53)
 *              %w      weekday (0-6, sunday is 0)
 *              %X      standard time representation
 *              %x      standard date representation
 *              %Y      year with century
 *              %y      year without century (00-99)
 *              %Z      timezone name
 *              %%      percent sign
 *
 *      the standard date string is equivalent to:
 *
 *          %a %b %d %Y
 *
 *      the standard time string is equivalent to:
 *
 *          %H:%M:%S
 *
 *      the standard date and time string is equivalent to:
 *
 *          %a %b %d %H:%M:%S %Y
 *
 *      strftime returns the number of characters placed in the
 *      buffer, not including the terminating \0, or zero if more
 *      than maxs characters were produced.
 *
**/

size_t strftime(char *s, size_t maxs, const char *f, const struct tm *t) {
    int w;
    char *p, *q, *r;

    char *buf = GETGCCCRAB()->strfbuf;

    p = s;
    q = s + maxs - 1;
    while ((*f != '\0')) {
        if (*f++ == '%') {
            r = buf;
            switch (*f++) {
                case '%' :
                    r = "%";
                    break;

                case 'a' :
                    r = aday[t->tm_wday];
                    break;

                case 'A' :
                    r = day[t->tm_wday];
                    break;

                case 'b' :
                    r = amonth[t->tm_mon];
                    break;

                case 'B' :
                    r = month[t->tm_mon];
                    break;

                case 'c' :
                    strfmt(r, "%0 %0 %2 %2:%2:%2 %4",
                           aday[t->tm_wday], amonth[t->tm_mon],
                           t->tm_mday, t->tm_hour, t->tm_min,
                           t->tm_sec, t->tm_year + 1900);
                    break;

                case 'd' :
                    strfmt(r, "%2", t->tm_mday);
                    break;

                case 'H' :
                    strfmt(r, "%2", t->tm_hour);
                    break;

                case 'I' :
                    strfmt(r, "%2", (t->tm_hour % 12) ? t->tm_hour % 12 : 12);
                    break;

                case 'j' :
                    strfmt(r, "%3", t->tm_yday + 1);
                    break;

                case 'm' :
                    strfmt(r, "%2", t->tm_mon + 1);
                    break;

                case 'M' :
                    strfmt(r, "%2", t->tm_min);
                    break;

                case 'p' :
                    r = (t->tm_hour > 11) ? "PM" : "AM";
                    break;

                case 'S' :
                    strfmt(r, "%2", t->tm_sec);
                    break;

                case 'U' :
                    w = t->tm_yday / 7;
                    if (t->tm_yday % 7 > t->tm_wday)
                        w++;
                    strfmt(r, "%2", w);
                    break;

                case 'W' :
                    w = t->tm_yday / 7;
                    if (t->tm_yday % 7 > (t->tm_wday + 6) % 7)
                        w++;
                    strfmt(r, "%2", w);
                    break;

                case 'w' :
                    strfmt(r, "%1", t->tm_wday);
                    break;

                case 'x' :
                    strfmt(r, "%3s %3s %2 %4", aday[t->tm_wday],
                           amonth[t->tm_mon], t->tm_mday, t->tm_year + 1900);
                    break;

                case 'X' :
                    strfmt(r, "%2:%2:%2", t->tm_hour,
                           t->tm_min, t->tm_sec);
                    break;

                case 'y' :
                    strfmt(r, "%2", t->tm_year % 100);
                    break;

                case 'Y' :
                    strfmt(r, "%4", t->tm_year + 1900);
                    break;

                case 'Z' :
                    r = (t->tm_isdst) ? __tzname[1] : __tzname[0];
                    break;

                default:
                    buf[0] = '%';     /* reconstruct the format */
                    buf[1] = f[-1];
                    buf[2] = '\0';
                    if (buf[1] == 0)
                        f--;        /* back up if at end of string */
            }
            while (*r) {
                if (p == q) {
                    *q = '\0';
                    return 0;
                }
                *p++ = *r++;
            }
        } else {
            if (p == q) {
                *q = '\0';
                return 0;
            }
            *p++ = f[-1];
        }
    }
    *p = '\0';
    return (size_t) (p - s);
}

static const int pow[5] = {1, 10, 100, 1000, 10000};

/**
 * static void strfmt(char *str, char *fmt);
 *
 * simple sprintf for strftime
 *
 * each format descriptor is of the form %n
 * where n goes from zero to four
 *
 * 0    -- string %s
 * 1..4 -- int %?.?d
 *
**/

static void strfmt(char *str, const char *fmt, ...) {
    int ival, ilen;
    char *sval;
    va_list vp;

    va_start(vp, fmt);
    while (*fmt) {
        if (*fmt++ == '%') {
            ilen = *fmt++ - '0';
            if (ilen == 0)                /* zero means string arg */
            {
                sval = va_arg(vp, char*);
                while (*sval)
                    *str++ = *sval++;
            } else                          /* always leading zeros */
            {
                ival = va_arg(vp, int);
                while (ilen) {
                    ival %= pow[ilen--];
                    *str++ = (char) ('0' + ival / pow[ilen]);
                }
            }
        } else *str++ = fmt[-1];
    }
    *str = '\0';
    va_end(vp);
}
